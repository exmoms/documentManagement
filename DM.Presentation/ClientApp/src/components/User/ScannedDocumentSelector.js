import Button from "@material-ui/core/Button";
import Card from "@material-ui/core/Card";
import CardActions from "@material-ui/core/CardActions";
import CardContent from "@material-ui/core/CardContent";
import CardMedia from "@material-ui/core/CardMedia";
import IconButton from "@material-ui/core/IconButton";
import { withStyles } from "@material-ui/core/styles";
import Typography from "@material-ui/core/Typography";
import DeleteForeverIcon from "@material-ui/icons/DeleteForever";
import MuiAlert from "@material-ui/lab/Alert";
import React, { Component } from "react";
import { withTranslation } from "react-i18next";
import CREATE_DOC_STYLES from "./Styles";
const style_ = CREATE_DOC_STYLES;

function Alert(props) {
  return <MuiAlert elevation={6} variant="filled" {...props} />;
}

class ScannedDocumentSelector extends Component {
  constructor(props) {
    super(props);

    this.state = {
      scannedDoc: [],
      scannedDocFiles: this.props.scannedDocFiles,
    };
    this.scannedDocInput = React.createRef();
    this.updateScannedFiles = this.props.updateScannedFiles;
  }
  t = this.props["t"];

  componentDidUpdate(prevProps) {
    if (prevProps.scannedDocFiles !== this.props.scannedDocFiles) {
      let scannedDocs_ = [];
      /* Map each file to a promise that resolves to an array of image URI's */
      Promise.all(
        this.props.scannedDocFiles.map((file) => {
          return new Promise((resolve, reject) => {
            const reader = new FileReader();
            reader.addEventListener("load", (ev) => {
              resolve(ev.target.result);
            });
            reader.addEventListener("error", reject);
            reader.readAsDataURL(file);
          });
        })
      ).then(
        (images) => {
          /* Once all promises are resolved, update state with image URI array */
          images.forEach((item) => scannedDocs_.push(item));
          this.setState({
            scannedDoc: scannedDocs_,
            scannedDocFiles: this.props.scannedDocFiles,
          });
        },
        (error) => {
          console.error(error);
        }
      );
    }
  }

  uploadScannedDoc = (event) => {
    // handle selected file to upload later
    var fileUploader = this.scannedDocInput;
    fileUploader.click();
  };

  onChangeFile = (event) => {
    if (event.target.files) {
      /* Get files in array form */
      const files = Array.from(event.target.files);

      /* save as files to post in submission process*/
      var tempFiles = new Set(this.state.scannedDocFiles);
      files.forEach((item) => tempFiles.add(item));

      /* Map each file to a promise that resolves to an array of image URI's */
      Promise.all(
        files.map((file) => {
          return new Promise((resolve, reject) => {
            const reader = new FileReader();
            reader.addEventListener("load", (ev) => {
              resolve(ev.target.result);
            });
            reader.addEventListener("error", reject);
            reader.readAsDataURL(file);
          });
        })
      ).then(
        (images) => {
          /* Once all promises are resolved, update state with image URI array */
          var temp = new Set(this.state.scannedDoc);
          images.forEach((item) => temp.add(item));
          this.setState({
            scannedDoc: Array.from(temp),
          });
          this.updateScannedFiles(Array.from(tempFiles), null);
        },
        (error) => {
          console.error(error);
        }
      );
      /* the following statement is import to allow re-select the last removed image */
      event.target.value = null;
    }
  };

  removePage = (event, index) => {
    event.stopPropagation();
    event.preventDefault();
    const { scannedDoc } = this.state;
    scannedDoc.splice(index, 1);

    this.setState({
      scannedDoc,
    });
    this.updateScannedFiles(null, index);
  };

  render() {
    const { classes } = this.props;
    const { scannedDoc } = this.state;
    return (
      <div style={{ padding: "20px 5px" }}>
        <input
          type="file"
          id="upload-scanned-doc-input"
          required
          accept="image/*"
          ref={(ref) => (this.scannedDocInput = ref)}
          style={{ display: "none" }}
          onChange={this.onChangeFile}
          multiple
        />
        <Card>
          <CardContent>
            <Typography gutterBottom variant="h5" component="h2" align="left">
              {this.t("previewdoc_scanneddoc")}
            </Typography>
            {this.props.error && (
              <Alert severity="error">{this.t("upload_error")}</Alert>
            )}
          </CardContent>
          <React.Fragment>
            {/* @todo don't display this element until the image is selected */}
            {/*       the scanned doc may contain multi-page, so we have multi-img */}
            {/* multi-page may need a button to add new part in case user didn't use multiple selection */}
            {scannedDoc.length !== 0 && (
              <div className={classes.buttonUploadScannedDoc}>
                {scannedDoc.map((imageURI, index) => (
                  <div className={classes.docMediaContainer} key={index}>
                    <IconButton
                      id={"IconButton" + index}
                      onClick={(event) => {
                        this.removePage(event, index);
                      }}
                      className={classes.deleteImgBtn}
                    >
                      <DeleteForeverIcon />
                    </IconButton>
                    <CardMedia
                      component="img"
                      image={imageURI}
                      className={classes.media}
                      id="scanned-doc-img-card"
                      title={this.t("previewdoc_uploadcopy")}
                    />
                  </div>
                ))}
              </div>
            )}
          </React.Fragment>
          <CardActions className={classes.buttonUploadScannedDoc}>
            <Button
              id="uploadScannedDoc"
              onClick={this.uploadScannedDoc}
              color="primary"
              size="small"
            >
              {this.t("previewdoc_uploadacopy")}
            </Button>
          </CardActions>
        </Card>
      </div>
    );
  }
}

export default withStyles(style_)(withTranslation()(ScannedDocumentSelector));
