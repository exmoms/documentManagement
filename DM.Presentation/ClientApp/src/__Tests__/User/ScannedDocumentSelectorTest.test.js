import React from "react";
import ReactDom from "react-dom";
import ScannedDocumentSelector from "../../components/User/ScannedDocumentSelector";
import { cleanup } from "@testing-library/react";
import renderer from "react-test-renderer";
import { act } from "react-dom/test-utils";
import { mount } from "enzyme";
import TextField from "@material-ui/core/TextField";
import CardMedia from "@material-ui/core/CardMedia";
import Card from "@material-ui/core/Card";
import CardActions from "@material-ui/core/CardActions";
import CardContent from "@material-ui/core/CardContent";
import DeleteForeverIcon from "@material-ui/icons/DeleteForever";
afterEach(cleanup);
describe("ScannedDocumentSelector", () => { 
  afterEach(() => {
    jest.clearAllMocks();
  });
  test("renders without crashing", () => {
    let updateScannedFilesMock = jest.fn();

    const div = document.createElement("div");
    ReactDom.render(
      <ScannedDocumentSelector
        updateScannedFiles={updateScannedFilesMock}
      ></ScannedDocumentSelector>,
      div
    );
  });

  test("ScannedDocumentSelector matches snapshot", () => {
    let updateScannedFilesMock = jest.fn();
  
    const component = renderer.create(
      <ScannedDocumentSelector updateScannedFiles={updateScannedFilesMock} />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("ScannedDocumentSelector, onChange functionality",  () =>{
    const updateScannedFilesMock = jest.fn();
    let mockValue = React.createRef();
    const mockSuccessResponse = ["image11" , "image22"];
    const mockJsonPromise = Promise.resolve(
      mockSuccessResponse
    );
    const PromiseMock = jest.spyOn(Promise, "all").mockImplementation(() => mockJsonPromise);
    const fileContents       = 'file contents';
    let wrapper = mount(
      <ScannedDocumentSelector updateScannedFiles={updateScannedFilesMock} />
    );

    return mockJsonPromise.then(() => {
      wrapper
      .find("#upload-scanned-doc-input")
      .at(0)
      .props()
      .onChange({ target: { files: [ new Blob([fileContents], {type : 'text/plain'}) ] , value : mockValue } });
    }).then(() => {
      wrapper.render();
      wrapper.update();
      //expect the side effects 
    const files = Array.from([ new Blob([fileContents], {type : 'text/plain'}) ]);
    var tempFiles = new Set();
    files.forEach((item) => tempFiles.add(item));
    expect(mockValue).toStrictEqual({"current": null});
    expect(PromiseMock).toHaveBeenCalledTimes(1);
      expect(updateScannedFilesMock).toHaveBeenCalledTimes(1);
      expect(updateScannedFilesMock.mock.calls[0][0]).toStrictEqual(Array.from(tempFiles));
    expect(updateScannedFilesMock.mock.calls[0][1]).toBe(null);
    expect(wrapper.find(CardMedia).length).toBe(2);
    });

  });

  test("ScannedDocumentSelector, onChange functionality when FileReader is a mock", () => {
    const updateScannedFilesMock = jest.fn();
    const fileContents       = 'file contents';
    const expectedFinalState = {fileContents: fileContents};
    const file               = new Blob([fileContents], {type : 'text/plain'});
    const addEventListener   = jest.fn();
    const readAsDataURL         = jest.fn();
    const dummyFileReader    = {addEventListener, readAsDataURL};
    window.FileReader        = jest.fn(() => dummyFileReader);
    let wrapper = mount(
      <ScannedDocumentSelector updateScannedFiles={updateScannedFilesMock} />
    );
    act(() => {
      wrapper
        .find("#upload-scanned-doc-input")
        .at(0)
        .props()
        .onChange({ target: { files: [ new Blob([fileContents], {type : 'text/plain'}) ] } });
    });
    wrapper.update();

    expect(addEventListener).toHaveBeenCalledTimes(2);
      expect(addEventListener.mock.calls[0][0]).toBe("load");
      expect(addEventListener.mock.calls[1][0]).toBe("error");
    expect(readAsDataURL).toHaveBeenCalledTimes(1);
    expect(readAsDataURL.mock.calls[0][0]).toStrictEqual(new Blob([fileContents], {type : 'text/plain'}));
    //expect the side effects 
    expect(updateScannedFilesMock).toHaveBeenCalledTimes(0);
    expect(wrapper.find(CardMedia).length).toBe(0);
  });

  test("ScannedDocumentSelector, removePage functionality",  () =>{
    const updateScannedFilesMock = jest.fn();
    const mockSuccessResponse = ["image11" , "image22"];
    const mockJsonPromise = Promise.resolve(
      mockSuccessResponse
    );
    const PromiseMock = jest.spyOn(Promise, "all").mockImplementation(() => mockJsonPromise);
    const fileContents       = 'file contents';
    let wrapper = mount(
      <ScannedDocumentSelector updateScannedFiles={updateScannedFilesMock} />
    );

    return mockJsonPromise.then(() => {
      wrapper
      .find("#upload-scanned-doc-input")
      .at(0)
      .props()
      .onChange({ target: { files: [ new Blob([fileContents], {type : 'text/plain'}) ] } });
    }).then(() => {
      wrapper.render();
      wrapper.update();
      //expect the side effects 
    const files = Array.from([ new Blob([fileContents], {type : 'text/plain'}) ]);
    var tempFiles = new Set();
    files.forEach((item) => tempFiles.add(item));
      expect(updateScannedFilesMock).toHaveBeenCalledTimes(1);
      expect(updateScannedFilesMock.mock.calls[0][0]).toStrictEqual(Array.from(tempFiles));
    expect(updateScannedFilesMock.mock.calls[0][1]).toBe(null);
    expect(wrapper.find(CardMedia).length).toBe(2);

    //click removeDutton 
    const stopPropagationMock = jest.fn();
    const preventDefaultMock = jest.fn();
    act(() => {
      wrapper
        .find("#IconButton0")
        .at(0)
        .props()
        .onClick({
          target: { id: 1, files: [{ name: "file1" }] },
          stopPropagation: stopPropagationMock,
          preventDefault: preventDefaultMock,
        });
    });
    wrapper.update();
    //expectations
    expect(stopPropagationMock).toHaveBeenCalledTimes(1);
    expect(preventDefaultMock).toHaveBeenCalledTimes(1);
    var temp = new Set([]);
    mockSuccessResponse.forEach((item) => temp.add(item));
    let scannedDoc =Array.from(temp);
    let valid = true;
    scannedDoc.splice(0, 1);
    if (scannedDoc.length === 0) {
      valid = false;
    }
    //test the update on scannedDoc
    expect(wrapper.find(CardMedia).length).toBe(1);
   //test the clalling for
   expect(updateScannedFilesMock).toHaveBeenCalledTimes(2);
    expect(updateScannedFilesMock.mock.calls[1][0]).toBe(null);
    expect(updateScannedFilesMock.mock.calls[1][1]).toBe(0);
    });
    
  });
});
afterEach(cleanup);
describe("ScannedDocumentSelector", () => { 
  afterEach(() => {
    jest.clearAllMocks();
  });
//TBD
// test("uploadScannedDoc, onClick for scannedDocInputRef functionality", () => {
//   const updateScannedFilesMock = jest.fn();
//   let wrapper = mount(
//     <ScannedDocumentSelector updateScannedFiles={updateScannedFilesMock} />
//   );
//   const clickInputSpy = jest.spyOn(HTMLInputElement.prototype, 'click');
//   act(() => {
//     wrapper
//       .find("#uploadScannedDoc")
//       .at(0)
//       .props()
//       .onClick({ target: { value: "" } });
//   });
//   wrapper.update();
//   expect(clickInputSpy).toHaveBeenCalledTimes(1);
// });
});