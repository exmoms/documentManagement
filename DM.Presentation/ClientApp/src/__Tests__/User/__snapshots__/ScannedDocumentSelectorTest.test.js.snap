// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScannedDocumentSelector ScannedDocumentSelector matches snapshot 1`] = `
<div
  style={
    Object {
      "padding": "20px 5px",
    }
  }
>
  <input
    accept="image/*"
    id="upload-scanned-doc-input"
    multiple={true}
    onChange={[Function]}
    required={true}
    style={
      Object {
        "display": "none",
      }
    }
    type="file"
  />
  <div
    className="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      className="MuiCardContent-root"
    >
      <h2
        className="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom MuiTypography-alignLeft"
      >
        previewdoc_scanneddoc
      </h2>
    </div>
    <div
      className="MuiCardActions-root withI18nextTranslation(ScannedDocumentSelector)-buttonUploadScannedDoc-15 MuiCardActions-spacing"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-textSizeSmall MuiButton-sizeSmall"
        disabled={false}
        id="uploadScannedDoc"
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          previewdoc_uploadacopy
        </span>
      </button>
    </div>
  </div>
</div>
`;
